File: backend/models/transaction.py
"""Transaction model."""
from datetime import datetime

from database.db import db


class Transaction(db.Model):
    __tablename__ = "transactions"

    id = db.Column(db.Integer, primary_key=True)
    raw_message = db.Column(db.String(1000), nullable=False)
    description = db.Column(db.String(1000), nullable=False)
    merchant = db.Column(db.String(255), nullable=False, default="Unknown Merchant")
    amount = db.Column(db.Float, nullable=False, default=0.0)
    transaction_type = db.Column(db.String(20), nullable=False, default="Unknown")  # Income | Expense | Unknown
    category = db.Column(db.String(50), nullable=False, default="Miscellaneous")
    is_manual_category = db.Column(db.Boolean, nullable=False, default=False)
    cashback_amount = db.Column(db.Float, nullable=True)
    reward_type = db.Column(db.String(50), nullable=True)  # Cashback | Reward Points | SuperCoins | Coins
    created_at = db.Column(db.DateTime, nullable=False, default=datetime.utcnow)
    updated_at = db.Column(db.DateTime, nullable=False, default=datetime.utcnow, onupdate=datetime.utcnow)

    def to_dict(self):
        cashback = None
        if self.cashback_amount is not None:
            cashback = {
                "rewardType": self.reward_type,
                "amount": self.cashback_amount,
            }

        return {
            "id": self.id,
            "description": self.description,
            "merchant": self.merchant,
            "amount": self.amount,
            "transactionType": self.transaction_type,
            "category": self.category,
            "isManualCategory": self.is_manual_category,
            "cashback": cashback,
            "createdAt": self.created_at.isoformat(),
            "updatedAt": self.updated_at.isoformat(),
        }

File: frontend/src/types/transaction.ts
export type TransactionType = "Income" | "Expense" | "Unknown";

export type Category =
  | "Food"
  | "Travel"
  | "Shopping"
  | "Bills"
  | "Entertainment"
  | "Salary"
  | "Medical"
  | "Miscellaneous";

export const CATEGORY_OPTIONS: Category[] = [
  "Food",
  "Travel",
  "Shopping",
  "Entertainment",
  "Salary",
  "Medical",
  "Bills",
  "Miscellaneous",
];

export interface Cashback {
  rewardType: string;
  amount: number | null;
}

export interface Transaction {
  id: number;
  description: string;
  merchant: string;
  amount: number;
  transactionType: TransactionType;
  category: Category;
  isManualCategory: boolean;
  cashback: Cashback | null;
  createdAt: string;
  updatedAt: string;
}

export interface DashboardSummary {
  totalIncome: number;
  totalExpense: number;
  balance: number;
  totalCashback: number;
  categoryTotals: Record<string, number>;
}

export interface CategorySlice {
  category: string;
  amount: number;
  percentage: number;
}

export interface AnalyticsSummary {
  pieChartData: CategorySlice[];
  progressBarData: CategorySlice[];
  topSpendingCategory: string | null;
  highestExpenseTransaction: Transaction | null;
  averageTransaction: number;
}

export type DateFilter = "today" | "7days" | "30days" | null;
export type TypeFilter = "Income" | "Expense" | null;
export type SortOption = "newest" | "oldest" | "highest" | "lowest" | "merchant";

export interface TransactionFilters {
  dateFilter: DateFilter;
  typeFilter: TypeFilter;
  categoryFilter: string | null;
  search: string;
  sortBy: SortOption;
}
